boards/BMS/Src/App/states/App_PreChargeState.c

```diff
@@ -20,11 +35,100 @@ static void
     PreChargeStateRunOnTick100Hz(struct StateMachine *const state_machine)
 {
     App_AllStatesRunOnTick100Hz(state_machine);
+
+    struct BmsWorld *   world = App_SharedStateMachine_GetWorld(state_machine);
+    struct Clock *      clock = App_BmsWorld_GetClock(world);
+    struct Airs *       airs  = App_BmsWorld_GetAirs(world);
+    struct Accumulator *accumulator = App_BmsWorld_GetAccumulator(world);
+    struct TractiveSystem *tractive_system =
+        App_BmsWorld_GetTractiveSystem(world);
+    struct PreChargeSequence *pre_charge_sequence =
+        App_BmsWorld_GetPreChargeSequence(world);
+    struct Charger *charger = App_BmsWorld_GetCharger(world);
+
+    // lower and upper bounds on expected charge time. Used to recognize faults
+    const uint32_t lower_precharge_time =
+        App_PreChargeSequence_GetLowerTimeBound_MS(
+            pre_charge_sequence,
+        App_TractiveSystem_GetSystemCapacitance());
+    const uint32_t upper_precharge_time =
+        App_PreChargeSequence_GetUpperTimeBound_MS(
+            pre_charge_sequence,
+            App_TractiveSystem_GetSystemCapacitance());
+
+    // Voltage information
+    float ts_voltage = App_TractiveSystem_GetSystemVoltage(tractive_system);
+    float threshold_voltage =
+        App_Accumulator_GetPackVoltage(accumulator) *
+        App_PreChargeSequence_GetThreshold_Percent(pre_charge_sequence);
+
+    // Time since precharge started
+    uint32_t elapsed_time =
+        App_SharedClock_GetCurrentTimeInMilliseconds(clock) -
+        App_SharedClock_GetPreviousTimeInMilliseconds(clock);
+
+    // CHARGE MODE
+    if (App_Charger_IsConnected(charger))
```

Comment:
should avoid repeating code here. should do precharge first regardless of whether the charger is connected then decide which state to go to next based on whether precharge has been completed